### DISCUSSION TITLE
Text Positions

### DISCUSSION THREAD

**Discussion by meghanaviyyapu** [pymupdf/PyMuPDF#1802] [Looking for help] [answered]
I have extracted rect values of text positions from a PDF. I want to insert text based on the positions extracted. Can you let me know if text matrix(Tm) values of a PDF can be obtained based on rect values? Is there any difference between rect values of text and text matrix values?

**Comment by JorjMcKie** [comment, marked as answer]
The text matrix `Tm` in PDF `/Contents` source is not directly available. Its use is also not intended in PyMuPDF - although you could of course read the `/Contents` as a stream and dig your way through to all information you want.

The bbox accompanying extracted text **_always_** already contains any computations required to deliver the correct value.

So you can use those bboxes (plus the `span["origin"]` - which is even more important) for inserting the span text.

**Reply by meghanaviyyapu** [reply]
Ok Thanks. Is there any way to compute text matrix values based on bbox values?

**Reply by JorjMcKie** [reply]
No. And even if, it would not (always) help you, because you may have multiple additional matrices in front upon which the Tm is based.

**Reply by JorjMcKie** [reply]
And as I said: what do need the Tm for in the first place?

**Reply by JorjMcKie** [reply]
If you absolutely want the lowest level of information, catch the XML output of `mutool trace your.pdf` and analyze it with package `lxml`.

**Reply by JorjMcKie** [reply]
mutool trace will give you information like this:
![grafik](https://user-images.githubusercontent.com/8290722/177796041-b6f5df75-5683-4435-8f5f-656a75530984.png)

**Reply by meghanaviyyapu** [reply]
I have extracted (left, right, top, bottom) values of table data cells. I wanted to insert the data using a different library in the content stream and to insert data I found a method called setTextMatrix. I could not figure out how to obtain those values during extraction and if insertion is possible using only text matrix values or if the data can be inserted based on (left, right, top, bottom) values.

**Reply by JorjMcKie** [reply]
In this case simple set the required text matrix to the identity matrix (1,0,0,1,0,0).
But confirm that that other package uses the same geometry as MuPDF: top-left = (0,0).
Otherwise use the matrix `~page.transformation_matrix` to recover PDF geometry: bottom-left = (0,0).

**Reply by meghanaviyyapu** [reply]
My objective is to insert text in a new PDF based on the text extracted from the input PDF. To insert text at the same position as that of input PDF, can you let me know the values that will be required and if extracted values can be used directly without any transformation?

**Reply by meghanaviyyapu** [reply]
> In this case simple set the required text matrix to the identity matrix (1,0,0,1,0,0). But confirm that that other package uses the same geometry as MuPDF: top-left = (0,0). Otherwise use the matrix `~page.transformation_matrix` to recover PDF geometry: bottom-left = (0,0).

Can you let me know what (x,y) values I should provide while inserting text to insert at the exact position of the input PDF?

**Reply by JorjMcKie** [reply]
This is `span["origin"]`. As I said: potentially to be multiplied with the transformation matrix inverse, depending on the geometry priciple of that other package.

**Reply by meghanaviyyapu** [reply]
Ok, Thanks for clarifying.
